module/validation/receipt_validator.go

```diff
@@ -167,10 +172,46 @@ func (v *receiptValidator) resultChainCheck(result *flow.ExecutionResult, prevRe
 //	* chunks are in correct format
 // 	* execution result has a valid parent
 // Returns nil if all checks passed successfully
-func (v *receiptValidator) Validate(receipt *flow.ExecutionReceipt) error {
+func (v *receiptValidator) Validate(receipts []*flow.ExecutionReceipt) error {
+	// lookup cache to avoid linear search when checking for previous result that is
+	// part of payload
+	payloadExecutionResults := make(map[flow.Identifier]*flow.ExecutionResult)
+	for _, receipt := range receipts {
+		payloadExecutionResults[receipt.ExecutionResult.ID()] = &receipt.ExecutionResult
+	}
+	// Build a functor that performs lookup first in receipts that were passed as payload and only then in
+	// local storage. This is needed to handle a case when same block payload contains receipts that
+	// reference each other.
+	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
+	// Eventually invalid result will be detected and fail the whole validation.
+	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+		if found {
+			return prevResult, nil
+		}
+
+		return v.previousResult(executionResult)
+	}
```

Comment:
⚠️ 
I am wondering if there is still an open edge case here:
* lets assume that the primary includes a receipt whose parent result has _not_ been incorporated in the respective fork. 
* However, if the replica validating the block already knows the receipt, it won't catch that protocol violation (?)

Reply:
a potential fix might be to include this logic into the mutator

Reply:
Here we get previous result from two places: 
1) Payload of current block. (`payloadExecutionResults`)
2) Storage (`v.previousResult`)
We can be sure that both of those execution results are valid, since if we get it from storage it must be valid. On other hand if we get it from payload, we actually don't know at this moment if parent is valid or not. But it will get checked eventually because we check every receipt in payload and fail the whole validation if one of them is invalid.
Am I missing something?

Reply:
I agree, that validity is guaranteed. 
However, in addition, we also have to check that there are no _gaps_ in the included receipts. I think the latter might not be guaranteed.

Reply:
> get previous result from Storage (v.previousResult)

Specifically that is the problem. The storage layer might know valid receipts, which have not been incorporated in the fork. However, validity of the fork should _not_ depend on any external information. Only the receipts incorporated in the fork should be used to validate the fork.

Reply:
I don't see why that is a problem, we store receipt into storage in two places, in `onReceipt` and when storing payload of block that is begin connected to the chain. In both cases we check with `receiptValidator` if that execution result is built on top of previous execution result, that way can be sure that they form a valid chain on the fork. I don't see how there can be a gap in results. Can you come up with an example where this is possible?

Reply:
Just discussed this with Leo. I am still convinced we are vulnerable to a byzantine proposer deliberately leaving out a receipt. Some consensus nodes will accept this block others can't if they don't have knowledge about the missing receipt

Reply:
but as we don't have byzantine consensus nodes at the moment, we can merge :-)

Reply:
Here is @AlexHentschel 's concern:

Let's say we have this chain, and we are building block X on top of C
```
A <- B(A) <- C <- X(C)
  ^- D (A) <- E(B)
```

X has a receipt C, it's an invalid block, but the validator would consider it as valid. It's should be invalid because the fork, where X is on, doesn't have result B. And result B is actually on a different fork, so it should be invalid. However, we consider it as valid, because we only require the parent result exists in the storage.

This is not a problem for matching engine, because the matching engine will actually go through all the fork, and ensure only the really incorporated results to be added. 

This will not be a problem for mutator, because Alex is solving that in mutator's logic.

Another problem is that if there is another consensus node, who don't see the fork `A <- D(A) <- E(B)`, then it won't have the result B in the storage, and it will consider X as invalid, which is a different conclusion than the first consensus node. In general, it's not good to have different conclusion when validating the same entity. 

Both problems could be solved by letting the receipt validator fork aware, the fork aware logic is currently implemented in both mutator and matching engine, so we might just need to extract them and move them to the receipt validator, and we should be good.

For now, we can leave it for later

Reply:
see issue https://github.com/dapperlabs/flow-go/issues/5253 for more explanation and proposal how to close this edge case.